rust/src/linearview.rs

```diff
@@ -0,0 +1,515 @@
+        let mut count: usize = 0;
+
+        let lines = unsafe { BNGetLinearViewCursorLines(self.handle, &mut count) };
+        LinearViewLinesIterator::new(count, lines)
+    }
+}
+
+impl PartialEq for LinearViewCursor {
+    fn eq(&self, other: &Self) -> bool {
+        unsafe { BNCompareLinearViewCursors(self.handle, other.handle) == 0 }
+    }
+}
+
+impl PartialOrd for LinearViewCursor {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        match unsafe { BNCompareLinearViewCursors(self.handle, other.handle) } {
+            i if i < 0 => Some(std::cmp::Ordering::Less),
+            i if i > 0 => Some(std::cmp::Ordering::Greater),
+            _ => Some(std::cmp::Ordering::Equal),
+        }
+    }
+}
+
+impl Ord for LinearViewCursor {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match unsafe { BNCompareLinearViewCursors(self.handle, other.handle) } {
+            i if i < 0 => std::cmp::Ordering::Less,
+            i if i > 0 => std::cmp::Ordering::Greater,
+            _ => std::cmp::Ordering::Equal,
+        }
+    }
+}
+
+unsafe impl RefCountable for LinearViewCursor {
+    unsafe fn inc_ref(handle: &Self) -> Ref<Self> {
+        Ref::new(Self {
+            handle: BNNewLinearViewCursorReference(handle.handle),
+        })
+    }
+
+    unsafe fn dec_ref(handle: &Self) {
+        BNFreeLinearViewCursor(handle.handle);
+    }
+}
+
+impl ToOwned for LinearViewCursor {
+    type Owned = Ref<Self>;
+
+    fn to_owned(&self) -> Self::Owned {
+        unsafe { RefCountable::inc_ref(self) }
+    }
+}
+
+unsafe impl Send for LinearViewCursor {}
+unsafe impl Sync for LinearViewCursor {}
+
+pub struct LinearViewLinesIterator {
+    count: usize,
+    lines: *mut BNLinearDisassemblyLine,
+    offset: usize,
+}
+
+impl LinearViewLinesIterator {
+    pub(crate) fn new(count: usize, lines: *mut BNLinearDisassemblyLine) -> Self {
+        Self {
+            count,
+            lines,
+            offset: 0,
+        }
+    }
+
+    pub(crate) fn empty() -> Self {
+        Self {
+            count: 0,
+            lines: std::ptr::null_mut(),
+            offset: 0,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.count
+    }
+}
+
+impl Iterator for LinearViewLinesIterator {
+    type Item = LinearDisassemblyLine;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset >= self.count {
+            return None;
+        }
+        let line = unsafe { LinearDisassemblyLine::from_raw(self.lines.offset(self.offset as _)) };
+
+        self.offset += 1;
+        Some(line)
+    }
+}
+
+
+// FIXME: we get a segmentation fault if this is there. However,
+// I don't understand enough of the API to see where this should get freed otherwise.
+/*
+impl Drop for LinearViewLinesIterator {
+    fn drop(&mut self) {
+         unsafe {
+             BNFreeLinearDisassemblyLines(self.lines, self.count);
+        }
+    }
+}
+*/
```

Comment:
I think this should all be replaceable with a `CoreOwnedArrayWrapper`.  That does the fancy `Guard` stuff and should deallocate correctly.

Reply:
I think the problem is in the when `LinearDisassemblyLine` gets dropped. I'll look at how `CoreOwnedArrayWrapper` works!

Reply:
The allocation issue persists also with `CoreOwnedArrayWrapper`, see 0d66f7b0910792107b60e3c2de550ebfca6f38eb

Reply:
Okay, after doing some digging, I believe the issue is this:
* `BNFreeLinearDisassemblyLines` will iterate over all lines, and call `BNFreeInstructionText` on each line's text.
* We have an `impl CoreOwnedArrayProvider for LinearDisassemblyLine` which (or in the case of the iterator, an `impl Drop`), in its `free` (or `drop`, which calls `free`) method calls `BNFreeLinearDisassemblyLines`, the lines will be freed.
* We have an `impl CoreOwnedArrayProvider for InstructionTextToken` calls `BNFreeInstructionText(raw, count);`.
* When we create an Array over the tokens in a Line, and that array gets dropped, the tokens are already freed.
* When we create an Array over the lines, and that array gets dropped, the lines will be attempted to be freed again, causing a double free.

I can think of the following solutions to this off the top of my head:
* Don't free the tokens in `<InstructionTextToken as CoreOwnedArrayProvider>::free`. This is not a nice option, since there are other scenarios where we would want an `Array<InstructionTextToken>`.
* Extend `CoreOwnedArrayProvider` to take a generic to implement context-sensitive resource deallocation. 
* Use an Iterator over the `InstructionTextToken` but an `Array<LinearDisassemblyLine>`.

Reply:
This sounds like an error in the bindings surrounding the InstructionTextToken array then; if data is being handed out of a Array<LinearDisassemblyLine> we shouldn't also be constructing an Array<InstructionTextToken> out of its contents, as Array<> objects have ownership semantics. If a LinearDisassemblyLine itself owns its array of InstructionTextTokens we can't be placing that interior array into an object that will free them out from under the actual owner when it drops.

Reply:
This might necessitate the quick creation of a `ArrayGuard<T>` path which is mostly equivalent to `Array` in containing ptr/count/ctxt but without the freeing behavior, similar to `Guard<T>` vs `Ref<T>`.

I'll double check to be sure, it's been several years since I wrote it originally.

Reply:
8be0b1a4f1e774b397df51b0bac265104f88972e adds an `ArrayGuard` for non-owned arrays by splitting `CoreOwnedArrayProvider` up.